### ISSUE TITLE
F/A-18C defineToggleSwitchToggleOnly2 sends incorrect performClickableAction 

### ISSUE THREAD

**Issue by kbastronomics** [DCSFlightpanels/dcs-bios#144]
In trying to resolved some issues with MAG switches pn the F/A-18C I identified one issue with the defineToggleSwitchToggleOnly2
On line 129 this should send a 0 not a 1 which prevents the switch from returning to the off position

Current code is this and issue is in the  fromState == 1 and toState == "0" part

```
	moduleBeingDefined.inputProcessors[msg] = function(toState)
		local fromState = GetDevice(0):get_argument_value(arg_number)
		if fromState == 0 and toState == "1" then
			GetDevice(device_id):performClickableAction(command, 1)
		end
		if fromState == 1 and toState == "0" then
			GetDevice(device_id):performClickableAction(command, 1)
		end
	end
```

this should be changed to

```
	moduleBeingDefined.inputProcessors[msg] = function(toState)
		local fromState = GetDevice(0):get_argument_value(arg_number)
		if fromState == 0 and toState == "1" then
			GetDevice(device_id):performClickableAction(command, 1)
		end
		if fromState == 1 and toState == "0" then
			GetDevice(device_id):performClickableAction(command, 0)
		end
	end
```

**Comment by WarLord211** [comment]
will change this.

**Comment by charliefoxtwo** [comment]
Nice catch!

**Comment by DeadMeatDCS** [comment]
@WarLord211 @kbastronomics a simpler alternative is to get rid of the special class altogether and use the defineToggleSwitch instead for 2-position mag switches (like it is for F16 Bit switch, a 2-pos mag as well)

Example code attached here working for APU (+ crank) if the defineToggleSwitch is used for APU.

[MagSwitchExamples_Hornet.zip](https://github.com/DCSFlightpanels/dcs-bios/files/10573035/MagSwitchExamples_Hornet.zip)

The example code could be cleaned/simplified I'm sure, but it works. Shared on OH thread previously.

**Comment by WarLord211** [comment]
i test this and if function is the same i change it

**Comment by WarLord211** [comment]
yesterday i didnt have much time to check it.
via refpage i cant get the switch back to 0 after the change from @kbastronomics 
also not with a simple toggle.
problem is the plane need a 1 (high) signal on the command to toggle

before a better solution is there i will withdraw the change

**Comment by kbastronomics** [comment]
> @WarLord211 @kbastronomics a simpler alternative is to get rid of the special class altogether and use the defineToggleSwitch instead for 2-position mag switches (like it is for F16 Bit switch, a 2-pos mag as well)
> 
> Example code attached here working for APU (+ crank) if the defineToggleSwitch is used for APU.
> 
> [MagSwitchExamples_Hornet.zip](https://github.com/DCSFlightpanels/dcs-bios/files/10573035/MagSwitchExamples_Hornet.zip)
> 
> The example code could be cleaned/simplified I'm sure, but it works. Shared on OH thread previously.

The issue I have with this approach is your basically bypass DCS-BIOS and coding switch control yourself for a good part of the switches actions,   I think we need a solution in fixing DCS-BIOS code so a user can use the supplied API.


I think all these tests and example help figure out the root cause and maybe if I was better with lua I could make more sense of what's happening.   the one thing I'm sure of it's its not in the arduino code since the reference page and bort shows same issue when sending switch actions to DCS

**Comment by DeadMeatDCS** [comment]
@WarLord211 That's strange. Using BORT I can toggle the APU switch just fine using both set state and INC/DEC commands to 0 OFF and 1 ON.

Did you change the switch definition to this?
`defineToggleSwitch("APU_CONTROL_SW", 12, 3001, 375, "Auxiliary Power Unit Panel", "APU Control Switch, ON/OFF")

Interestingly, sending

**Comment by DeadMeatDCS** [comment]
@kbastronomics My contribution here could be considered interim I guess, in the sense that it works but could be better..
As soon as a true native solution is working it should take its place. 

There are examples of instruments/controls outside the Hornet like the IFF panel on A10 and others that requires some **creativity** in the sketch to make them work in a realistic way with hardware. There's some included in the library for the M2000 radio presets for example.

**Comment by kbastronomics** [comment]
@DeadMeatDCS understand.   My intent on OH code is to remain as pure DCS-BIOS as possible to make it easier on our user base.  One of my requirements I was given.

**Comment by DeadMeatDCS** [comment]
@kbastronomics the principles I've shared in the working sketch could most likely be wrapped into a lua class to make it easier on the user. That'd be killer. I just don't have the skills to do it myself.

See for example Switches.h where there's also debouncing, pin reading and sendDcsBiosMessages going on. 
Only real new contribution here is the state trick we're using to unblock. Beyond that I don't think there's anything that is not "pure" - it is simply unwieldy for the user at present because it is done in sketch directly.

But it works (in my testing at least). For me this is more important than it being super simple :)
That's why I propose the defineToggleSwitch change @WarLord211 (and @charliefoxtwo for that matter) as a change until something better is created (like a new mag switch lua class).

But fair enough I'm also completely fine with no change being made for now if it is too confusing. I mean there's the no less ugly HID workaround that's also working with the current module right ;)

**Comment by kbastronomics** [comment]
@DeadMeatDCS I would still content that it is not the solution we should be looking for.   it does not fix the root cause of the issue only does what we've done to help our user base move forward for that reason should not INHO be made part of DCS-BIOS code, especially since it's a temporary circumvention not a fix.

Finding the root cause of why DCS-BIOS is not processing that return action unless it's sent a fake one is the actual fix that I believe should be pursued.   adding a circumvention into DCS-BIOS code I believe is not the answer.

That of course is up to @WarLord211 , @charliefoxtwo

**Comment by WarLord211** [comment]
i will make some test as i created a new switch for the M1 maybe it work for this too

**Comment by charliefoxtwo** [comment]
It sounds like we're talking about two separate issues.

The way I'm reading it, @DeadMeatDCS is talking about removing the special casing for specific aircraft and condensing logic. On its face this appears to be a great suggestion, but it doesn't _seem_ material to this issue - at least to me. We should open a new issue to track that work.

Meanwhile, @kbastronomics is suggesting digging deeper to find why values set by the game logic in this module are ignored, which I believe is the heart of this specific issue. Sure, we might have a few different toggle switch implementations that should probably be condensed, but I don't know that any of them are _so_ different as to cause this - it definitely seems like something closer to the core of dcs-bios to me.

**Comment by kbastronomics** [comment]
@charliefoxtwo Thank you for your insight.  

Maybe that's been my sources of confusion for lack of a better word.
So we continue to track the issue with why values set by  game logic are being ignored here and @DeadMeatDCS opens a new issue for his item to address.

I think that is a good plan and helps us keep items seperate.

**Comment by WarLord211** [comment]
@DeadMeatDCS I testet now a switches from the class and no one work as toggle.
As Toggle the APU only work if INC is Used, not DEC or TOGGLE so i think this class is not the right one

**Comment by Gosling616** [comment]
Where has this issue (@kbastronomics 's) moved to, please?

**Comment by kbastronomics** [comment]
I understood this issue was to remain and @DeadMeatDCS was to open a new issue.
